### PR TITLE
Expand indirection cell calls in IR instead of doing ad-hoc optimizations in codegen

### LINKED ISSUE #65076: JIT: Control-flow guard checked interface calls should not materialize cell addresses twice
### Description

We currently materialize indirection cell addresses twice for CFG checked interface calls on arm64:

### Reproduction Steps

```powershell
$env:COMPlus_TieredCompilation=0
$env:COMPlus_JitForceControlFlowGuard=1
```

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static void M(I i)
{
    i.Foo();
}
```

We produce the following code:
```asm
G_M6547_IG02:              ;; offset=0008H
        D280010F          movz    x15, #8
        F2B9218F          movk    x15, #0xc90c LSL #16
        F2CFFFCF          movk    x15, #0x7ffe LSL #32
        F94001EF          ldr     x15, [x15]
        94000000          bl      CORINFO_HELP_VALIDATE_INDIRECT_CALL
        AA0F03E1          mov     x1, x15
        D280010B          movz    x11, #8
        F2B9218B          movk    x11, #0xc90c LSL #16
        F2CFFFCB          movk    x11, #0x7ffe LSL #32
        D63F0020          blr     x1
```

### Expected behavior

We should only load the indirection cell address once.

### Actual behavior

We load it twice.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by jakobbotsch**
When the JIT calls a function that requires an indirection cell argument
it is beneficial to load the call target through the already placed arg.
Previously this was done as an ad-hoc optimization during codegen (with
LSRA making sure to allocate an extra register). Instead of this, expand
it as regular IR and remove the ad-hoc optimization.

This change makes the optimization that were previously done work for
CFG as well and generally seems to have (albeit small) positive diffs.

Fix #65076

### CODE REVIEW

File: src/coreclr/jit/gentree.h [hunk 1]
```diff
@@ -4683,47 +4684,15 @@ struct GenTreeCall final : public GenTree
...
-        // On XARCH we disassemble it from callsite except for tailcalls that need indirection cell.
-        if (IsR2RRelativeIndir() && IsFastTailCall())
-        {
-            return NonStandardArgKind::R2RIndirectionCell;
-        }
-#endif
-
-        return NonStandardArgKind::None;
-    }
+    bool HasIndirectionCellArg(fgArgTabEntry** entry) const;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/gentree.h:4687]: Consider adding a default arg value so callers who don't care about the entry don't need to explicitly pass `nullptr`.

```suggestion
    bool HasIndirectionCellArg(fgArgTabEntry** entry = nullptr) const;
```

File: src/coreclr/jit/gentree.h [hunk 2]
```diff
@@ -4683,47 +4684,15 @@ struct GenTreeCall final : public GenTree
...
 
     CFGCallKind GetCFGCallKind()
     {
 #if defined(TARGET_AMD64)
         // On x64 the dispatcher is more performant, but we cannot use it when
         // we need to pass indirection cells as those go into registers that
         // are clobbered by the dispatch helper.
-        bool mayUseDispatcher    = GetIndirectionCellArgKind() == NonStandardArgKind::None;
+        bool mayUseDispatcher    = !HasIndirectionCellArg(nullptr);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/gentree.h:4695]: If you add the default as above:
```suggestion
        bool mayUseDispatcher    = !HasIndirectionCellArg();
```

File: src/coreclr/jit/valuenum.cpp [hunk 3]
```diff
@@ -9759,22 +9759,14 @@ void Compiler::fgValueNumberHelperCallFunc(GenTreeCall* call, VNFunc vnf, ValueN
         vnpUniq.SetBoth(vnStore->VNForExpr(compCurBB, call->TypeGet()));
     }
 
-#if defined(FEATURE_READYTORUN) && defined(TARGET_ARMARCH)
-    if (call->IsR2RRelativeIndir())
+    if (call->HasIndirectionCellArg(nullptr))
```
Comment (AndyAyersMS) [inline src/coreclr/jit/valuenum.cpp:9762]: And here
```suggestion
    if (call->HasIndirectionCellArg())
```

**Review by AndyAyersMS — APPROVED** [review]
LGTM.

Seems like arm is not yet happy though.

**Comment by jakobbotsch** [comment]
The ARM issue seemed to be an LSRA bug: for R2R calls we were not killing the register containing the indirection cell address at the call site which led to that register being marked busy for the remainder of the function. ARM is a bit special there since it uses a callee-preserved register. There was a hack in morph to work around this problem (I assume) that I have removed as well.

**Comment by jakobbotsch** [comment]
On x64 with CFG it looks like it's better to use the constant in both places:
```diff
 ; MyBenchmarks.SimpleIfaceCall.Call()
        push      rax
        xor       eax,eax
        mov       [rsp],rax
        mov       rcx,[rcx+8]
        mov       [rsp],rcx
-       mov       r10,7FF972ED0510
-       mov       rcx,[r10]
+       mov       rcx,[7FF9736E0510]
        call      CORINFO_HELP_VALIDATE_INDIRECT_CALL
        mov       rax,rcx
        mov       rcx,[rsp]
-       mov       r11,r10
+       mov       r11,7FF972EA0510
        cmp       [rcx],ecx
        add       rsp,8
        jmp       rax
-; Total bytes of code 52
+; Total bytes of code 53
 ```
we get
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1526 (21H2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=7.0.100-alpha.1.21558.2
  [Host]     : .NET 6.0.0 (6.0.21.48005), X64 RyuJIT
  Job-DBANTR : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT
  Job-BFNMJK : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT

EnvironmentVariables=COMPlus_JitForceControlFlowGuard=1  

```
| Method |        Job |                 Toolchain |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD | Code Size |
|------- |----------- |-------------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|
|   Call | Job-DBANTR |    \Core_Root\corerun.exe | 1.599 ns | 0.0778 ns | 0.2271 ns | 1.510 ns |  1.21 |    0.23 |      52 B |
|   Call | Job-BFNMJK | \Core_Root_am\corerun.exe | 1.332 ns | 0.0604 ns | 0.1602 ns | 1.288 ns |  1.00 |    0.00 |      53 B |

where the _am version is the diff version. Will disable the expansion with locals for x64 CFG calls with validator.

**Comment by jakobbotsch** [comment]
Not totally sure if this change can replace the old ad-hoc optimization. A few observations:
1. In large functions with many locals the new indir cell locals may not be tracked/may cause other locals to not be tracked, which lead to regressions. It will probably help to reuse the local.
2. In tier0/debug codegen the new locals cause extra spilling, which is to be expected. Not sure if this is really an issue (this is the cause of the aspnet regressions, which seem to be in tier0 method contexts when spot checking)
3. On ARM32 LSRA seems to be bad at dealing with the new locals (and new local when reusing it). Still have not been able to pinpoint one particular issue, it seems the new local leads to globally different register allocation which causes regressions.

**Comment by jakobbotsch** [comment]
Closing this, don't think we can teach RA/liveness to deal well with the extra locals required for this, and the regressions seem too large to ignore.

**Pull request closed without merging**